swarmrl/engine/espresso.py

```diff
@@ -630,36 +680,40 @@ def integrate(self, n_slices, force_model: swarmrl.models.InteractionModel = Non
                     for val in self.traj_holder.values():
                         val.clear()
 
-            swarmrl_colloids = []
-            if force_model is not None:
-                for col in self.colloids:
-                    swarmrl_colloids.append(
-                        swarmrl.models.interaction_model.Colloid(
-                            pos=col.pos, director=col.director, id=col.id, type=col.type
-                        )
-                    )
-                actions = force_model.calc_action(swarmrl_colloids)
-                for action, coll in zip(actions, self.colloids):
-                    coll.swimming = {"f_swim": action.force}
-                    coll.ext_torque = action.torque
-                    new_direction = action.new_direction
-                    if new_direction is not None:
-                        if self.n_dims == 3:
-                            coll.director = new_direction
-                        else:
-                            old_direction = coll.director
-                            rotation_angle = np.arccos(
-                                np.dot(new_direction, old_direction)
-                            )
-                            if rotation_angle > 1e-6:
-                                rotation_axis = np.cross(old_direction, new_direction)
-                                rotation_axis /= np.linalg.norm(rotation_axis)
-                                # only values of [0,0,1], [0,0,-1] can come out here,
-                                # plusminus numerical errors
-                                rotation_axis = [0, 0, round(rotation_axis[2])]
-                                coll.rotate(axis=rotation_axis, angle=rotation_angle)
-
-            self.system.integrator.run(self.params.steps_per_slice)
+            if (
+                self.step_idx * self.params.time_step.m_as("sim_time")
+                >= self.params.time_slice.m_as("sim_time") * self.slice_idx - 1e-6
+            ):
+                self.slice_idx += 1
+                self.manage_forces(force_model)
+
+            # print("write idx, slice idx, step idx",self.write_idx,
+            # self.slice_idx, self.step_idx)
+            time_to_next_write = (
```

Comment:
you can avoid a lot of this logic by requiring that write_interval and time_slice are multiples of each other (does not matter which one is the smaller one). You always integrate the smaller number and do the more frequent thing of [forces, writing] until you do the less frequent thing